qiskit/transpiler/passes/optimization/optimize_1q_gates.py

```diff
@@ -174,13 +185,21 @@ def run(self, dag):
                                                  (2 * np.pi)) == 0:
                     right_name = "nop"
 
+            if right_name == "u1" and "u1" not in self.basis:
+                right_name = "u3"
```

Comment:
Should we try `u2` first?

Reply:
I dont think it makes any difference... I made the change in 9b72b75f4. Can you come up with a test case?

Reply:
Sure, a oneQ gate which is determined to be a right `u1` should be the first of `u1`, `u2`, `u3` which is in `self.basis` (because a `u1` is cheaper to implement than a `u2` which is cheaper to implement than a `u3`).

The check could be with an `elif` but would have to be adapted to handle the "'u2' if 'u2' in self.basis else 'u3'" case. Personally, I think its okay as is.

Reply:
No this should be u3 here because a u1 cannot fall back to a u2. 

The name might suggest a u1 can be a u2 can be a u3, but in fact u1 --> u3 and u2 --> u3 but u1 -/-> u2.

```
        U1(\lambda) =
            \begin{pmatrix}
                1 & 0 \\
                0 & e^{i\lambda}
            \end{pmatrix}

        U2(\phi, \lambda) = \frac{1}{\sqrt{2}}
            \begin{pmatrix}
                1          & e^{-i\lambda} \\
                e^{i\phi} & e^{i(\phi+\lambda)}
            \end{pmatrix}
```